solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java

```diff
@@ -192,9 +215,106 @@ public ModelWeight createWeight(IndexSearcher searcher, boolean needsScores, flo
+      this.req = req;
+    }
+
+    public FeatureWeight call() throws Exception{
+      try {
+        FeatureWeight fw  = f.createWeight(searcher, needsScores, req, originalQuery, efi);
+        return fw;
+      } catch (final Exception se) {
+        throw new FeatureException("Exception from createWeight for " + f.toString() + " "
+            + se.getMessage(), se);
+      } finally {
+        LTRThreadInterface.ltrSemaphore.release();
+      }
+    }
+  } // end of call CreateWeightCallable
+
+  private ModelWeight createWeightsSelectivelyParallel(IndexSearcher searcher, boolean needsScores, float boost) throws IOException {
+
+    long  time1 = System.currentTimeMillis();
+    final Collection<Feature> modelFeatures = meta.getFeatures();
+    final Collection<Feature> allFeatures = meta.getAllFeatures();
+
+    int modelFeatSize = modelFeatures.size();
+    int allFeatSize = this.extractAllFeatures ? allFeatures.size() : modelFeatSize;
+
+    final FeatureWeight[] extractedFeatureWeights = new FeatureWeight[allFeatSize];
+    final FeatureWeight[] modelFeaturesWeights = new FeatureWeight[modelFeatSize];
+    final SolrQueryRequest req = getRequest();
+    //= req.getParams().getInt("LTRThreads", 1);
+    Executor executor = LTRThreadInterface.createWeightScoreExecutor;
+    if  (LTRThreadInterface.ltrSemaphore == null ){
+      LTRThreadInterface.ltrSemaphore = new Semaphore((LTRThreadInterface.maxThreads <=0) ? 10 : LTRThreadInterface.maxThreads);
+    }
+    Collection<Feature> features = null;
+    // since the feature store is a linkedhashmap order is preserved
+    if (this.extractAllFeatures) {
+      features = allFeatures;
+    }
+    else{
+      features =  modelFeatures;
+    }
+    List<Future<FeatureWeight> > futures = new ArrayList<>(features.size());
+    List<FeatureWeight > featureWeights = new ArrayList<>(features.size());
+    try{
+      for (final Feature f : features) {
+        CreateWeightCallable callable = new CreateWeightCallable(f, searcher, needsScores, req);
+        RunnableFuture<FeatureWeight> runnableFuture = new FutureTask<>(callable);
+        LTRThreadInterface.ltrSemaphore.acquire();//may block and/or interrupt
+        executor.execute(runnableFuture);//releases semaphore when done
+        futures.add(runnableFuture);
+      }
+      //Loop over futures to get the feature weight objects
+      for (final Future<FeatureWeight> future : futures) {
+        featureWeights.add(future.get());
+      }
+    } catch (InterruptedException e) {
+      log.info("Error while creating weights in LTR: InterruptedException", e);
+    } catch (ExecutionException ee) {
+      Throwable e = ee.getCause();//unwrap
+      if (e instanceof RuntimeException) {
+        throw (RuntimeException) e;
+      }
+      log.info("Error while creating weights in LTR: " + e.toString(), e);
+    }
+    long  time2 = System.currentTimeMillis();
+    int i=0, j = 0;
+    if (this.extractAllFeatures) {
```

Comment:
See if you can refactor both createWeight functions.  This bottom half assigning into the featureWeights arrays, and the top half creating the containers is pretty much exactly the same.